### ISSUE TITLE
Images don't display in knowledge posts.

### ISSUE THREAD

**Issue by sryza** [airbnb/knowledge-repo#155]
Auto-reviewers: @NiharikaRay @matthewwardrop @earthmancash @danfrankj

Hi all,

I've deployed the knowledge server on top of a knowledge repo that's a MySQL DB, and I'm having trouble getting images to display in any knowledge post.

I'm running on top of master, and I've verified that the images have entries in the knowledge repo database.  For lack of a better way that I'm aware of, I've been "publishing" posts by manually updating the status of the DB entries to 3.  So it's possible that's the source of the problem?

Any idea what could be going on?  Many thanks for any guidance.

**Comment by matthewwardrop** [comment]
Thanks for reporting this @sryza !

We'll look into it and see if we can reproduce it. If we can, we'll fix it; otherwise we'll try to work with you to see what's going on.

Thanks again!

**Comment by ajbosco** [comment]
@sryza What are the steps for deploying a knowledge repo that's a MySQL DB?

**Comment by matthewwardrop** [comment]
@ajbosco You can just do:

`knowledge_repo --repo mysql://<username>@<host>/<database>:<table> deploy`

**Comment by sryza** [comment]
@matthewwardrop if I wanted to look into this myself, do you have recommendations on where in the code would be a good starting point?

**Comment by earthmancash** [comment]
@sryza what does your post look like? And is it all being written in the webeditor, or are you starting from RMarkdown / iPython?

**Comment by sryza** [comment]
@earthmancash I haven't used the webeditor.  The post is an IPython post with a few matplotlib-generated figures.

**Comment by earthmancash** [comment]
Ok, so you're putting these iPython posts on a github repo, which the server box has locally when running the webapp? Can you tell me what the image link looks like in the knowledge.md on the git repo? Wondering if there's an issue with the path in there and where the image path is on the locally cloned git repo. (unless you're using an s3 like cloud service, in which case it's likely a different issue)

**Comment by sryza** [comment]
@earthmancash I'm using the database knowledge repo, so the knowledge.md file and the image are stored as separate rows in the database.  The ref column for the image row has the value "images/output_22_1.png".  The image is referenced inside the knowledge.md as  `![png](images/output_22_1.png)`.

Does that answer your question?

**Comment by earthmancash** [comment]
I think I understand. @matthewwardrop has more familiarity with what an iPython --> dbrepo setup should look like. Can you help Sandy out with where to look to debug the issue?

**Comment by matthewwardrop** [comment]
Hi @sryza,

Many apologies for the delay here... All manner of things have been vying for my attention over the last couple of weeks.

The general flow of image rendering in knowledge posts is governed by `KnowedgePostProcessor`s, which detects if a "local" image (such as the uri you have given as an example) exists in the knowledge post, and then replaces the uri with a base64 encoded image uri. I'm guessing something has gone wrong with the iteration of "files" stored in the database, so that the `KnowledgePost.image_paths` is not properly iterating over the images in the post.

Can you report the output of the following code (run in python script or interpreter)?

```
from knowledge_repo import KnowledgeRepository

repo = KnowledgeRepository('mysql://......')
post = repo.post('<path to post in repo>')

print(post.image_paths)
print(list(post.read_images().keys()))
```

**Comment by sryza** [comment]
Thanks for the guidance @matthewwardrop.  It looks like your suggestion is correct.  Both of those print statements print out empty lists.

However, if I print `list(post._dir())`, I get the following output:
```
['orig_src/pharmacy_incentive.ipynb',
 'images/output_12_1.png',
 'images/output_10_1.png',
 'images/output_12_2.png',
 'images/output_14_1.png',
 'images/output_22_1.png',
 'images/output_12_0.png',
 'knowledge.md']
```

It looks like the problem relates to the handling of the `parent` arg in `DbKnowledgeRepository#_kp_dir`.  That method prepends the post path to the parent, so, in my case, it queries the database for entries where path=`pharmacy_incentive/pharmacy_incentive.kp/images`.  However, this conflicts the way the entries are stored in the database:

```
| c38e1be0-78a6-44e9-8519-fade62c92228 | pharmacy_incentive/pharmacy_incentive.kp | images/output_12_1.png            |
| c38e1be0-78a6-44e9-8519-fade62c92228 | pharmacy_incentive/pharmacy_incentive.kp | images/output_10_1.png            |
| c38e1be0-78a6-44e9-8519-fade62c92228 | pharmacy_incentive/pharmacy_incentive.kp | images/output_12_2.png            |
| c38e1be0-78a6-44e9-8519-fade62c92228 | pharmacy_incentive/pharmacy_incentive.kp | images/output_14_1.png            |
```

I.e. "images/" goes under the `ref` column, not the `path` column.  So it seems like the solution depends on which column "images/" should go under?  If "images/" should go under `path`, then the solution is to make that change.  If "images/" should stay under `ref`, then the solution is to look for it there.

**Comment by matthewwardrop** [comment]
I suspect that is probably the case :). Thanks for looking into this. The `DBKnowledgeRepository` has been far less tested than the `GitKnowledgeRepository` backend.

Do you want to put up a PR fixing it, or do you want me to do that?

**Comment by matthewwardrop** [comment]
The way I'd imagine this working, btw, is to look up paths by joining the two fields; something like:

`SELECT data from posts where path|'/'|ref = {path}`

**Comment by sryza** [comment]
@matthewwardrop I'm happy to work on a PR.  So, if I understand correctly, you're saying that the "images/" prefix should stay within the `ref` column?

**Comment by matthewwardrop** [comment]
Yep. Posts are identified by their virtual path; and then there are named "references" for each post; which may include nested structure.

Looking forward to your PR.

**Comment by matthewwardrop** [comment]
Fixed by your patch in #171. Thanks again!
